supply/lib/supply/client.rb

```diff
@@ -5,9 +5,74 @@
 require 'net/http'
 
 module Supply
-  class Client
+  UI = FastlaneCore::UI unless defined?(UI)
+
+  class AbstractGoogleServiceClient
+    SCOPE = nil
+    SERVICE = nil
+
     # Connecting with Google
-    attr_accessor :android_publisher
+    attr_accessor :client
+
+    def self.make_from_config(params: nil)
+      unless params[:json_key] || params[:json_key_data]
+        UI.important("To not be asked about this value, you can specify it using 'json_key'")
+        params[:json_key] = UI.input("The service account json file used to authenticate with Google: ")
+      end
+
+      if params[:json_key]
+        service_account_json = File.open(File.expand_path(params[:json_key]))
+      elsif params[:json_key_data]
+        service_account_json = StringIO.new(params[:json_key_data])
+      end
+
+      return self.new(service_account_json: service_account_json, params: params)
+    end
+
+    # Initializes the service and its auth_client using the specified information
+    # @param service_account_json: The raw service account Json data
+    def initialize(service_account_json: nil, params: nil)
+      auth_client = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+
+      UI.verbose("Fetching a new access token from Google...")
+
+      auth_client.fetch_access_token!
+
+      if FastlaneCore::Env.truthy?("DEBUG")
+        Google::Apis.logger.level = Logger::DEBUG
+      end
+
+      Google::Apis::ClientOptions.default.application_name = "fastlane (supply client)"
+      Google::Apis::ClientOptions.default.application_version = Fastlane::VERSION
+      Google::Apis::ClientOptions.default.read_timeout_sec = params[:timeout]
+      Google::Apis::ClientOptions.default.open_timeout_sec = params[:timeout]
+      Google::Apis::ClientOptions.default.send_timeout_sec = params[:timeout]
+      Google::Apis::RequestOptions.default.retries = 5
+
+      service = self.class::SERVICE
+      service.authorization = auth_client
+
+      if params[:root_url]
+        # Google's client expects the root_url string to end with "/".
+        params[:root_url] << '/' unless params[:root_url].end_with?('/')
+        service.root_url = params[:root_url]
+      end
+
+      self.client = service
+    end
+
+    private
+
+    def call_google_api
+      yield if block_given?
+    rescue Google::Apis::ClientError => e
+      UI.user_error!("Google Api Error: #{e.message}")
+    end
+  end
+
+  class Client < AbstractGoogleServiceClient
+    SERVICE = Androidpublisher::AndroidPublisherService.new
```

Comment:
I don't know if we want to actually call `.new`here. There is a line in `AbstractGoogleServiceClient` that does the following...

```rb
service = self.class::SERVICE
service.authorization = auth_client
```

This will modify the class `SERVICE` class variable. I THINK what we want to do instead is change `SERVICE` to just be `SERVICE = Androidpublisher::AndroidPublisherService` and then create a new method in `AbstractGoogleServiceClient` that creates the instance like...

```rb
def service
  @service ||= self.class::self.class::SERVICE
  return @service
end
```

Or something like this? Let me know if my concern sounds crazy 🙃

Reply:
Oh right, that first line doesn't create a copy.

But you probably meant:
```
  @service ||= self.class::SERVICE.new
```
Correct?

We could also just do `service = self.class::SERVICE.new` possibly and not add `@service` into the mix?

Reply:
Depends 😊 `@service ||= self.class::SERVICE.new` has a benefit of the service only creating one single instance so that you can call the `service` method multiple times without needing to worry about accidentally creating multiple instances

Reply:
Ehhh, ish... `attr_accessor` automatically creates the `@variable_name` and the getter and setter methods that go along with it. This is a little bit different because this method is just a getter that is initializing the `@service` if it hasn't been initialized and then returning it